parcels/particle.py

```diff
@@ -290,55 +350,58 @@ def execute(self, pyfunc=AdvectionRK4, starttime=None, endtime=None, dt=1.,
 
-        # Check if starttime, endtime, runtime and dt are consistent and compute timesteps
+        # Derive starttime, endtime and interval from arguments or grid defaults
         if runtime is not None and endtime is not None:
             raise RuntimeError('Only one of (endtime, runtime) can be specified')
         if starttime is None:
-            if dt > 0:
-                starttime = self.grid.time[0]
-            else:
-                starttime = self.grid.time[-1]
+            starttime = self.grid.time[0] if dt > 0 else self.grid.time[-1]
         if runtime is not None:
-            if dt > 0:
-                endtime = starttime + runtime
-            else:
-                endtime = starttime - runtime
+            endtime = starttime + runtime if dt > 0 else starttime - runtime
         else:
             if endtime is None:
-                if dt > 0:
-                    endtime = self.grid.time[-1]
-                else:
-                    endtime = self.grid.time[0]
-        if endtime < starttime and dt > 0:
-            dt = -1. * dt
-            print("negating dt because running in time-backward mode")
-        if endtime > starttime and dt < 0:
-            dt = -1. * dt
-            print("negating dt because running in time-forward mode")
-        timesteps = int((endtime - starttime) / dt)
-
-        # Check if output is required and compute outer leaps
-        if output_interval <= 0 or output_interval > abs(endtime-starttime):
-            output_steps = timesteps
-        else:
-            output_steps = abs(int(output_interval / dt))
-        timeleaps = int(timesteps / output_steps)
-        # Execute kernel in sub-stepping intervals (leaps)
-        current = starttime
+                endtime = self.grid.time[-1] if dt > 0 else self.grid.time[0]
+        if interval is None:
+            interval = endtime - starttime
+
+        # Ensure that dt and interval have the correct sign
+        if endtime > starttime:  # Time-forward mode
+            if dt < 0:
+                dt *= -1.
+                print("negating dt because running in time-forward mode")
+            if interval < 0:
+                interval *= -1.
+                print("negating interval because running in time-forward mode")
+        if endtime < starttime:  # Time-backward mode
+            if dt > 0.:
+                dt *= -1.
+                print("negating dt because running in time-backward mode")
+            if interval > 0.:
+                interval *= -1.
+                print("negating interval because running in time-backward mode")
+
+        # Initialise particle timestepping
+        for p in self:
+            p.time = starttime
+            p.dt = dt
+        # Execute time loop in sub-steps (timeleaps)
+        timeleaps = int((endtime - starttime) / interval)
+        assert(timeleaps >= 0)
+        leaptime = starttime
         for _ in range(timeleaps):
```

Comment:
I also think it would be worth thinking about the particle density calculation here and how it goes in the loop. You'll see I had a go at this in PR #67. Perhaps you could take a look and include the relevant code from there and we can scrap that PR?